py/server/deephaven/table_factory.py

```diff
@@ -231,68 +232,48 @@ def write_row(self, *values: Any) -> None:
-    not permitted.
-
-    The keyed input tablet has keys for each row and supports addition/deletion/modification of rows by the keys.
-    """
+    def __init__(self, source_table: Table):
+        super().__init__(source_table)
+        self.j_input_table = self.j_table.getAttribute(_J_INPUT_TABLE_ATTRIBUTE)
+        if not self.j_input_table:
+            raise DHError("the provided table input is not suitable for input tables.")
 
-    def __init__(self, col_defs: Dict[str, DType] = None, init_table: Table = None,
-                 key_cols: Union[str, Sequence[str]] = None):
-        """Creates an InputTable instance from either column definitions or initial table. When key columns are
-        provided, the InputTable will be keyed, otherwise it will be append-only.
+    def add(self, table: Table) -> None:
+        """Synchronously writes rows from the provided table to this input table. If this is a keyed input table, added rows with keys
+        that match existing rows will replace those rows.
 
         Args:
-            col_defs (Dict[str, DType]): the column definitions
-            init_table (Table): the initial table
-            key_cols (Union[str, Sequence[str]): the name(s) of the key column(s)
+            table (Table): the table that provides the rows to write
 
         Raises:
             DHError
         """
         try:
-            if col_defs is None and init_table is None:
-                raise ValueError("either column definitions or init table should be provided.")
-            elif col_defs and init_table:
-                raise ValueError("both column definitions and init table are provided.")
-
-            if col_defs:
-                j_arg_1 = _JTableDefinition.of(
-                    [Column(name=n, data_type=t).j_column_definition for n, t in col_defs.items()])
-            else:
-                j_arg_1 = init_table.j_table
-
-            key_cols = to_sequence(key_cols)
-            if key_cols:
-                super().__init__(_JKeyedArrayBackedInputTable.make(j_arg_1, key_cols))
-            else:
-                super().__init__(_JAppendOnlyArrayBackedInputTable.make(j_arg_1))
-            self.j_input_table = self.j_table.getAttribute(_J_INPUT_TABLE_ATTRIBUTE)
-            self.key_columns = key_cols
+            self.j_input_table.add(table.j_table)
         except Exception as e:
-            raise DHError(e, "failed to create a InputTable.") from e
+            raise DHError(e, "add to InputTable failed.") from e
 
-    def add(self, table: Table) -> None:
-        """Writes rows from the provided table to this input table. If this is a keyed input table, added rows with keys
+    def add_async(self, table: Table, j_listener: jpy.JType = _JInputTableStatusListener.DEFAULT) -> None:
+        """Asynchronously  writes rows from the provided table to this input table. If this is a keyed input table, added rows with keys
```

Comment:
We generally don't expose raw JPY types in public methods. Please refer to the adapter pattern we use for wrapping Python functions/methods for table listeners in `table_listener.py`.